### PR TITLE
[SPARK-33824][PYTHON][DOCS] Restructure and improve Python package management page

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to restructure and refine the Python dependency management page.
I lately wrote a blog post which will be published soon, and decided contribute some of the contents back to PySpark documentation.
FWIW, it has been reviewed by some tech writers and engineers.

I built the site for making the review easier: https://hyukjin-spark.readthedocs.io/en/stable/user_guide/python_packaging.html

### Why are the changes needed?

For better documentation.

### Does this PR introduce _any_ user-facing change?

It's doc change but only in unreleased bracnhs for now.

### How was this patch tested?

I manually built the docs as:

```bash
cd python/docs
make clean html
open 
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @srowen, @zero323, @viirya, @dongjoon-hyun, @erikerlandson, would you mind taking a quick look when you guys find some time?

**Comment by HyukjinKwon** [comment]
retest this please

File: python/docs/source/user_guide/python_packaging.rst [hunk 1]
```diff
@@ -118,87 +116,139 @@ Example using ``SparkSession.builder``:
...
 .. code-block:: bash
 
-    pex pyspark==3.0.1 pyarrow==0.15.1 pandas==0.25.3 -o myarchive.pex
+    python -m venv pyspark_venv
+    source pyspark_venv/bin/activate
+    pip install pyarrow pandas venv-pack
+    venv-pack -o pyspark_venv.tar.gz
 
+You can directly pass/unpack the archive file and enable the environment on executors by leveraging
+the ``--archives`` option or ``spark.archives`` configuration (``spark.yarn.dist.archives`` in Yarn).
```
Comment (srowen) [inline python/docs/source/user_guide/python_packaging.rst:155]: Very minor but Yarn -> YARN

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]

File: python/docs/source/user_guide/python_packaging.rst [hunk 2]
```diff
@@ -51,65 +51,63 @@ Here is the script ``app.py`` from the previous example that will be executed on
...
 PySpark allows to upload Python files (``.py``), zipped Python packages (``.zip``), and Egg files (``.egg``)
-to the executors by setting the configuration setting ``spark.submit.pyFiles`` or by directly
-calling :meth:`pyspark.SparkContext.addPyFile`.
+to the executors by:
 
-This is an easy way to ship additional custom Python code to the cluster. You can just add individual files or zip whole
-packages and upload them. Using :meth:`pyspark.SparkContext.addPyFile` allows to upload code
-even after having started your job.
+- Setting the configuration setting ``spark.submit.pyFiles``
+- Ssing ``--py-files`` option in Spark scripts
```
Comment (viirya) [inline python/docs/source/user_guide/python_packaging.rst:69]: Ssing? Setting?

File: python/docs/source/user_guide/python_packaging.rst [hunk 3]
```diff
@@ -118,87 +116,139 @@ Example using ``SparkSession.builder``:
...
+.. code-block:: bash
+
+    ./pyspark_pex_env.pex -c "import pandas; print(pandas.__version__)"
+    1.1.5
+
+However, ``.pex`` file does not include a Python interpreter itself under the hood so all
+nodes in a cluster should have the same Python interpreter installed.
+
+In order to transfer and use the ``.pex`` file in a cluster, you should ship it via the
+``spark.files`` configuration (``spark.yarn.dist.files`` in Yarn) or ``--files`` option because they are regular files instead
```
Comment (viirya) [inline python/docs/source/user_guide/python_packaging.rst:215]: Yarn is here too.

**Review by viirya — APPROVED** [review]
lgtm with few minor comments.

**Review by zero323 — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Thank you guys. Merged to master and bracnh-3.1.

**Comment by HyukjinKwon** [comment]
Just in case anyone is interested in, my blog post was published yesterday here https://databricks.com/blog/2020/12/22/how-to-manage-python-dependencies-in-pyspark.html :-).

**Pull request closed without merging**